File: Prototype01.js
// class pessoa - padrao que sera clonada:

class Pessoa{
    constructor(id, nome, idade){
        this.id = id
        this.nome = nome;
        this.idade = idade;
    }

    // Metodo clone - realiza uma copia superficial do objeto:
    clone(){
        return new Pessoa(this.id, this.nome, this.idade);
    }

}

// classe PessoaManager - gerencia instancias de pessoa:
class PessoaManager{
    constructor(){
        this.pessoas = {};
    }

// metodo addPessoa - adiciona pessoas ao adicionar pessoa:
    addPessoa(id, nome, idade){
        const pessoa = new Pessoa(id, nome, idade);
        this.pessoas[id] = pessoa;
    }

    //metodo getPessoaId - solicita uma pessoa eplo id e retorna uma copia dela:
    getPessoaId(id){
        const pessoaOriginal = this.pessoas[id];
        if(pessoaOriginal){
            return pessoaOriginal.clone();
        }else{
            return null;
        }
    }
}

//criando uma instancia de PessoaManager:
    const manager = new PessoaManager();

// add duas pessoas:
manager.addPessoa(1, 'Joao da Silva', 30);
manager.addPessoa(2, 'Maria da Silva', 27);

//clonando 1 pessoa e modificando
const pessoaClone1 = manager.getPessoaId(1);
if(pessoaClone1){
    pessoaClone1.nome = 'Dolly da Silva Clonado';
    pessoaClone1.id = 3
}

// clonando 1 pessoa e modificando:
const pessoaClone2 = manager.getPessoaId(2);
if(pessoaClone2){
    pessoaClone2.idade = '9999';
    pessoaClone2.id = 4
}

//imprimindo pessoas:
console.log("Pessoa Original - 1:");
console.log(manager.getPessoaId(1));
console.log("Pessoa Clonada 1:");
console.log(pessoaClone1);
console.log("------------------------------------------------")
console.log("Pessoa Original - 2:");
console.log(manager.getPessoaId(2));
console.log("Pessoa Clonada 2:");
console.log(pessoaClone2);
console.log("------------------------------------------------")

File: Prototype01.py
import copy

# Classe Pessoa - Padrão que será clonada:
class Pessoa:
    def __init__(self, nome, idade):
        self.nome = nome
        self.idade = idade

    def __str__(self):
        return f"Nome: {self.nome}, Idade: {self.idade}"

    def clone(self):
        return copy.copy(self)
    
# Classe PessoaManager - Gerencia instancias de Pessoa:
class PessoaManager:
    def __init__(self):
        self.pessoas = {}
    
    def addPessoa(self, nome, idade, id):
        pessoa = Pessoa(nome, idade)
        self.pessoas[id] = pessoa
        
    def getPessoaById(self, id):
        return self.pessoas[id].clone()

# Criando uma instancia de PessoaManager:
manager = PessoaManager()

# Add duas pessoas:
manager.addPessoa('João da Silva', 30, 1);
manager.addPessoa('Maria da Silva', 27, 2);

# Clonando 1 Pessoas e Modificando:
pessoaClone1 = manager.getPessoaById(1)
pessoaClone1.idade = 18

# Clonando 2 Pessoas e Modificando:
pessoaClone2 = manager.getPessoaById(2)
pessoaClone2.nome = "Dolly da Silva Clonada"
pessoaClone2.idade = 55

# Imprimindo as Pessoas:
print("Pessoa Original - 1:")
print(manager.getPessoaById(1))
print("Pessoal Clonada 1:")
print(pessoaClone1)
print("--------------------------------------")
print("Pessoa Original - 2:")
print(manager.getPessoaById(2))
print("Pessoal Clonada 2:")
print(pessoaClone2)
print("--------------------------------------")